homeassistant/components/knx/__init__.py

```diff
@@ -245,39 +297,44 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
         schema=SERVICE_KNX_EXPOSURE_REGISTER_SCHEMA,
     )
 
-    async def reload_service_handler(service_call: ServiceCall) -> None:
-        """Remove all KNX components and load new ones from config."""
-
-        # First check for config file. If for some reason it is no longer there
-        # or knx is no longer mentioned, stop the reload.
-        config = await async_integration_yaml_config(hass, DOMAIN)
-        if not config or DOMAIN not in config:
-            return
+    return True
 
-        await asyncio.gather(
-            *(platform.async_reset() for platform in async_get_platforms(hass, DOMAIN))
-        )
-        await knx_module.xknx.stop()
 
-        await async_setup(hass, config)
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unloading the KNX platforms."""
+    knx_module: KNXModule = hass.data[DOMAIN]
+    for exposure in knx_module.exposures:
+        exposure.shutdown()
 
-    async_register_admin_service(
-        hass, DOMAIN, SERVICE_RELOAD, reload_service_handler, schema=vol.Schema({})
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        entry,
+        [
+            platform.value
+            for platform in SupportedPlatforms
+            if platform.value in hass.data[DATA_KNX_CONFIG]
+        ],
     )
+    if unload_ok:
+        await knx_module.stop()
+        hass.data[DOMAIN] = None
+        hass.data[DATA_KNX_CONFIG] = None
```

Comment:
Better to remove it so you get KeyError if some code is still active instead of passing around `None` and then failing at a random later moment.

```suggestion
        hass.data.pop(DOMAIN)
        hass.data.pop(DATA_KNX_CONFIG)
```